api/lql.go

```diff
@@ -18,24 +18,201 @@
+)
+
+type LQLQuery struct {
+	ID             string `json:"LQL_ID,omitempty"`
+	StartTimeRange string `json:"START_TIME_RANGE,omitempty"`
+	EndTimeRange   string `json:"END_TIME_RANGE,omitempty"`
+	QueryText      string `json:"QUERY_TEXT"`
+	// QueryBlob is a special string that supports type conversion
+	// back and forth from LQL to JSON
+	QueryBlob string `json:"-"`
+}
+
+func (q *LQLQuery) Validate(allowEmptyTimes bool) error {
+	// translate
+	if err := q.Translate(); err != nil {
+		return err
+	}
+	// validate range
+	if err := q.ValidateRange(allowEmptyTimes); err != nil {
+		return err
+	}
+	// validate query
+	if q.QueryText == "" {
+		return errors.New("query should not be empty")
+	}
+	return nil
+}
+
+func (q *LQLQuery) Translate() error {
+	// query
+	if err := q.TranslateQuery(); err != nil {
+		return err
+	}
+	// start
+	if start, err := q.TranslateTime(q.StartTimeRange); err != nil {
+		return err
+	} else {
+		q.StartTimeRange = start
+	}
+	// end
+	if end, err := q.TranslateTime(q.EndTimeRange); err != nil {
+		return nil
+	} else {
+		q.EndTimeRange = end
+	}
+	return nil
+}
+
+func (q *LQLQuery) TranslateQuery() error {
+	// empty
+	if q.QueryText != "" {
+		return nil
+	}
+	// json
+	var t LQLQuery
+
+	if err := json.Unmarshal([]byte(q.QueryBlob), &t); err == nil {
+		if q.StartTimeRange == "" {
+			q.StartTimeRange = t.StartTimeRange
+		}
+		if q.EndTimeRange == "" {
+			q.EndTimeRange = t.EndTimeRange
+		}
+		q.QueryText = t.QueryText
+		return err
+	}
+	// lql
+	if matched, _ := regexp.MatchString(reLQL, q.QueryBlob); matched {
+		q.QueryText = q.QueryBlob
+		return nil
+	}
+	return errors.New(LQLQueryTranslateError)
+}
+
+func (q LQLQuery) TranslateTime(inTime string) (outTime string, err error) {
+	// empty
+	if inTime == "" {
+		return
+	}
+	// parse time as RFC3339
+	if t, err := time.Parse(time.RFC3339, inTime); err == nil {
+		outTime = t.UTC().Format(time.RFC3339)
+		return outTime, err
+	}
+	// parse time as millis
+	if t, err := strconv.ParseInt(inTime, 10, 64); err == nil {
+		outTime = time.Unix(0, t*int64(time.Millisecond)).UTC().Format(time.RFC3339)
+		return outTime, err
+	}
+	err = errors.New("unable to parse time (" + inTime + ")")
+	return
+}
+
+func (q LQLQuery) ValidateRange(allowEmptyTimes bool) (err error) {
+	// validate start
+	var start time.Time
+	if q.StartTimeRange != "" {
+		if start, err = time.Parse(time.RFC3339, q.StartTimeRange); err != nil {
+			return err
+		}
+	} else if allowEmptyTimes {
+		start = time.Unix(0, 0)
+	} else {
+		return errors.New("start time must not be empty")
+	}
+	// validate end
+	var end time.Time
+	if q.EndTimeRange != "" {
+		if end, err = time.Parse(time.RFC3339, q.EndTimeRange); err != nil {
+			return err
+		}
+	} else if allowEmptyTimes {
+		end = time.Now()
+	} else {
+		return errors.New("end time must not be empty")
+	}
+	// validate range
+	if start.After(end) {
+		return errors.New("date range should have a start time before the end time")
```

Comment:
apologies but if you are using "named returns" this is an antipattern, either use them or don't. 😅
```suggestion
		err = errors.New("date range should have a start time before the end time")
		return
```

Reply:
Fixed.